### PR TITLE
Add company address fields

### PR DESCRIPTION
**Pull request by marcofucci**
### Description of change

This:
- adds `address_*` fields to the model
- adds `address_*` fields to the admin with a message explaining why 3 addresses exist
- changes the company API to populate `address_*` fields from trading address if defined or registered address otherwise
- improves some existing company tests

screenshot 2019-01-22 at 11 29 15

### Checklist

* [x] Has a new newsfragment been created? Check [changelog/README.rst](https://github.com/uktrade/data-hub-leeloo/blob/master/changelog/README.rst) for instructions
* [ ] Have any relevant search models been updated?
* [ ] Have any relevant fixtures (`fixtures/test_data.yaml`) been updated?
* [ ] Have any relevant select-/prefetch-related field lists in the views and search apps been updated?
* [x] Has the admin site been updated (for new models, fields etc.)?
* [ ] Has the README been updated (if needed)?

### CODE REVIEW

File: datahub/company/models/company.py [hunk 1]
```diff
@@ -168,6 +168,24 @@ class Company(ArchivableModel, BaseModel, CompanyAbstract):
...
+    address_country = models.ForeignKey(
+        metadata_models.Country,
+        blank=True,
+        null=True,
+        on_delete=models.PROTECT,
+        related_name='company_address_country',
+    )
+    address_postcode = models.CharField(max_length=MAX_LENGTH, blank=True, null=True)
+
+    # will eventually become obsolete when the migration to solely address and registred
```
Comment (elcct) [inline datahub/company/models/company.py]: A typo `registred`.

File: datahub/company/serializers.py [hunk 2]
```diff
@@ -382,8 +403,52 @@ def validate(self, data):
...
                 })
 
+        self._populate_address_fields(combiner, data)
+
         return data
 
+    def _populate_address_fields(self, combiner, data):
+        """
+        Populates the address_* fields in data with values from trading address if
+        defined or registered address otherwise.
```
Comment (elcct) [inline datahub/company/serializers.py]: I would reword this comment, I find it difficult to read. 

Maybe something like: Populates the address_* fields with the values from trading address or registered address whichever is defined.

File: datahub/company/serializers.py [hunk 3]
```diff
@@ -382,8 +403,52 @@ def validate(self, data):
...
+                mapping.values()
+                for mapping in CompanySerializer.ADDRESS_FIELDS_MAPPING.values()
+            ],
+        )
+        was_any_address_field_specified = any(
+            field_name in data
+            for field_name in all_address_field_names
+        )
+
+        # if no address field was passed in, don't do anything
```
Comment (elcct) [inline datahub/company/serializers.py]: This is probably not necessary as I think the code below is self-explanatory.

File: datahub/company/models/company.py [hunk 4]
```diff
@@ -168,6 +168,24 @@ class Company(ArchivableModel, BaseModel, CompanyAbstract):
         metadata_models.UKRegion, blank=True, null=True,
         on_delete=models.SET_NULL,
     )
+
+    # address is the main location for the business, it could be the trading address
+    # or the registered address or a completely different address
+    address_1 = models.CharField(max_length=MAX_LENGTH, blank=True, null=True)
```
Comment (reupen) [inline datahub/company/models/company.py]: Do we want `null=True` on the new char fields?

File: datahub/company/test/factories.py [hunk 5]
```diff
@@ -41,12 +41,22 @@ class CompanyFactory(factory.django.DjangoModelFactory):
         factory.Faker('company'),
         factory.Faker('company'),
     ])
+
     registered_address_1 = factory.Sequence(lambda n: f'{n} Foo st.')
     registered_address_town = 'London'
+    registered_address_postcode = factory.Faker('postcode', locale='en_GB')
```
Comment (reupen) [inline datahub/company/test/factories.py]: We default to `en_GB` in `datahub/conftest.py`, do we need to specify it again here?

File: datahub/company/models/company.py (hunk 4 above)
Comment (marcofucci) [reply datahub/company/models/company.py]: yeah, otherwise we can't do 0-downtime development, right?

File: datahub/company/serializers.py (hunk 3 above)
Comment (marcofucci) [reply datahub/company/serializers.py]: 👍

File: datahub/company/models/company.py (hunk 4 above)
Comment (reupen) [reply datahub/company/models/company.py]: Oh, yes, that's true.

File: datahub/company/serializers.py [hunk 6]
```diff
@@ -382,8 +403,52 @@ def validate(self, data):
...
+
+        TODO: delete after the migration to address and registered address is completed
+        """
+        all_address_field_names = itertools.chain(
+            *[
+                mapping.values()
+                for mapping in CompanySerializer.ADDRESS_FIELDS_MAPPING.values()
+            ],
+        )
+        was_any_address_field_specified = any(
```
[the remark this answers is not in the document]
Comment (marcofucci) [reply datahub/company/serializers.py]: Sure, if you think it makes it more readable to you

File: datahub/company/models/company.py (hunk 4 above)
Comment (marcofucci) [reply datahub/company/models/company.py]: I'll add some TODOs though

Comment (marcofucci) [reply datahub/company/models/company.py]: and defaults

File: datahub/company/test/factories.py (hunk 5 above)
Comment (marcofucci) [reply datahub/company/test/factories.py]: Forgot about that, thanks!

File: datahub/company/migrations/0062_add_address.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    dependencies = [
+        ('metadata', '0021_delete_companyclassification'),
+        ('company', '0061_remove_alias_from_database'),
+    ]
+
+    operations = [
+        migrations.AddField(
+            model_name='company',
+            name='address_1',
+            field=models.CharField(blank=True, default='', max_length=255, null=True),
```
[the remark this answers is not in the document]
Comment (marcofucci) [reply datahub/company/migrations/0062_add_address.py]: Oh damn, forgot that, so magic! I'll remove the defaults then.
